Fixed punycoding of URLs and emails.

Brandon-T opened the pull request:
This fixes the ability to search with quoted queries.
This fixes the ability to search with email address as a search query.
Fixes https://github.com/brave/brave-ios/issues/1209

Security Review: https://github.com/brave/internal/issues/664

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

<!-- Any useful notes for reviewer explaining how best to test and verify. -->

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [ ] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [ ] Issues include necessary QA labels:
  - [ ] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [ ] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable)

Brandon-T committed:
Added documentation.

kylehickinson commented on Client/Frontend/Browser/URIFixup.swift:
```diff
@@ -32,6 +32,18 @@ class URIFixup {
         }
+        
+        // A URL is only valid when the URL has a scheme, is not an email, is not quoted, and can be punycoded.
+        // If one of the above conditions is NOT satisfied, the URL is invalid and should be deemed "search terms" search terms instead. Technically, an email is also a valid URL but does not get handled by the DNS server.
```
Can we fix-up these comments to be contained to a max 110 char column count?

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/URIFixup.swift:
```diff
@@ -50,4 +62,18 @@ class URIFixup {
+            "]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-" +
+            "9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21" +
+            "-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"
+        return NSPredicate(format: "SELF MATCHES[c] %@", regex).evaluate(with: string)
```
Interesting, never thought of using `NSPredicate` here instead of `NSRegularExpression`, is there any diff?

Brandon-T replied:
Hmm I never thought to use any of the other options.. however, it seems that `range(of:)` is actually the most optimized between all three with -O.. It doesn't let me specify -O3 but I'm sure with O3 it'd be negligible or produce the same results.

https://godbolt.org/z/p8XcFE

I'll switch to `range(of:)` which is more legible anyway.

Brandon-T replied:
Fixed.

iccub commented on Client/Frontend/Browser/URIFixup.swift:
```diff
@@ -50,4 +62,18 @@ class URIFixup {
+    
+    /// Checks whether a string is a valid email conforming to RFC 2822.
+    /// http://www.cocoawithlove.com/2009/06/verifying-that-string-is-email-address.html
+    static func isValidEmail(_ string: String) -> Bool {
```
Could we add few unit tests to it? That regex looks scary :D

Brandon-T replied:
Fixed. Added Punycode unit tests.

Brandon-T replied:
Added Unit tests for emails. Although, that regex conforms to RFC2822 strictly, I added on that allows more "relaxed" testing which omits quotes and square brackets which still conforms to RFC2822.

Brandon-T committed:
Changed isValidEmail to guard against empty and to use range(of:) instead of predicates.

Brandon-T committed:
Added Unit tests and a more relaxed Regex for modern day emails based on RFC2822.

danishjafri88 reviewed (commented):
Sorry for not mentioning this earlier, I forgot my security fix for eTLD had the `punycode` implementation removed from the app. The URI fixes are now handled by Webkit hence no longer required to be done in the app. Please refer to https://github.com/brave/brave-ios/pull/869 `Client/Frontend/Browser/URIFixup.swift` changes

Brandon-T commented:
> Sorry for not mentioning this earlier, I forgot my security fix for eTLD had the `punycode` implementation removed from the app. The URI fixes are now handled by Webkit hence no longer required to be done in the app. Please refer to #869 `Client/Frontend/Browser/URIFixup.swift` changes

Hmm.. that pull request has been open for 6 months? Also, when you remove PunyCoding, does the `homograph` attack become effective? or?

EDIT: Just tested. The homograph attack is mitigated in your PR still.. However, it does NOT fix the bug this PR fixes (Entering Emails or Quoted strings: #1209).

danishjafri88 commented:
> > Sorry for not mentioning this earlier, I forgot my security fix for eTLD had the `punycode` implementation removed from the app. The URI fixes are now handled by Webkit hence no longer required to be done in the app. Please refer to #869 `Client/Frontend/Browser/URIFixup.swift` changes
> 
> Hmm.. that pull request has been open for 6 months? Also, when you remove PunyCoding, does the `homograph` attack become effective? or?

Webkit now ensures that IDN are now Punycoded by default. try `ΓΝΩΘΙΣΕΑΥΤΟΝ.EXAMPLE.COM` as an example.

danishjafri88 commented:
> mitigated

Agreed, we need your fix for email, just making sure that we don't use the old `punycoded` method anymore.

Brandon-T commented:
> > mitigated
> 
> Agreed, we need your fix for email, just making sure that we don't use the old `punycoded` method anymore.

We will need your PR merged first then. After that, I will have to redo this PR and apply the fix.

iccub reviewed (approved):
Code looks good, thanks.

Could you add some Test Plan for QA please?

jhreis commented:
I apologize for #869 falling through the cracks. It was a lengthy discussed PR and had many moving parts, but should have been merged long ago. It is now merged.

Brandon-T committed:
Fixed punycoding of URLs and emails.
This fixes the ability to search with quoted queries.
This fixes the ability to search with email address as a search query.
Fixes https://github.com/brave/brave-ios/issues/1209

Brandon-T committed:
Added Unit Test which discovered quite a bit of bugs in the punycoding of unvalidated URLs.

Brandon-T committed:
Added email unit tests.

Brandon-T committed:
Much more simplified fix!

Brandon-T commented:
> I apologize for #869 falling through the cracks. It was a lengthy discussed PR and had many moving parts, but should have been merged long ago. It is now merged.

:)

Updated the PR to have a much simpler fix now. No more need for checking if it's an email specifically via a Regex. I took some of the code from the PR with the safe-browsing and applied it here.

Specifically "canonicalization" and then checking if the url has a "user" for a "domain" :D. This fixed a hidden bug as well in Chrome that Safari handles perfectly!

iccub commented:
@Brandon-T  Please submit a security ticket if it hasn't been done already

diracdeltas commented on Client/Frontend/Browser/URIFixup.swift:
```diff
@@ -6,6 +6,23 @@ import Foundation
+        // Validate the domain to make sure it doesn't have any invalid characters
+        // IE: quotes, etc..
+        if let host = url.host {
+            guard let decodedASCIIURL = host.replacingOccurrences(of: "+", with: "").removingPercentEncoding else {
```
why is `+` removed?

Brandon-T replied:
IIRC, it was for email searching as emails could have a plus sign in the host but the function would have returned invalid.. but the code has changed since I first wrote it, and it works without it.

Should be fixed now.

diracdeltas commented on ClientTests/SearchTests.swift:
```diff
@@ -45,6 +45,20 @@ class SearchTests: XCTestCase {
+        checkInvalidURL("\"foo@brave.com")
+        checkInvalidURL("\"foo@brave.com\"")
+        checkInvalidURL(#""创业咖啡.中国"#)
+        checkInvalidURL(#""创业咖啡.中国""#)
```
thx for adding tests

Brandon-T committed:
Removed plus sign removal from host (legacy code).

diracdeltas commented:
lgtm
